src/go/serviceconfig/service_config_fetcher.go

```diff
@@ -61,57 +61,91 @@ func NewServiceConfigFetcher(mf *metadata.MetadataFetcher, opts options.ConfigGe
+
+			return "", serviceConfig, nil
 		}
 
-		glog.Infof("check new rollouts for service %v", scf.serviceName)
-		newRolloutId, newConfigId, err := scf.loadConfigFromRollouts(scf.serviceName, scf.curRolloutId, scf.curConfigId())
+		glog.Infof("check new config id for service %v", s.serviceName)
+		newRolloutId, err := s.newRolloutId()
 		if err != nil {
-			glog.Errorf("error occurred when checking new rollouts, %v", err)
+			return "", nil, fmt.Errorf("error occurred when checking new service rollout id: %v", err)
 		}
-		if scf.curRolloutId != newRolloutId && scf.curConfigId() != newConfigId {
-			scf.curRolloutId = newRolloutId
-			token, _, err := scf.accessToken()
-			if err != nil {
-				return nil, fmt.Errorf("fail to get access token: %v", err)
-			}
-			return scf.callServiceManagement(util.FetchConfigURL(scf.opts.ServiceManagementURL, scf.serviceName, newConfigId), token)
+
+		if newRolloutId == s.curRolloutId {
+			return "", nil, nil
+		}
+
+		rollout := new(smpb.Rollout)
+		fetchRolloutUrl := util.FetchRolloutURL(s.opts.ServiceManagementURL, s.serviceName, newRolloutId)
+		if err = util.CallGooglelapis(s.client, fetchRolloutUrl, util.GET, s.accessToken, rollout); err != nil {
+			return "", nil, err
+		}
+
+		newConfigId, err := highestTrafficConfigIdInRollout(rollout)
+		if err != nil {
+			return "", nil, err
+		}
+
+		if newConfigId == s.curConfigId() {
+			return newRolloutId, nil, nil
+		}
+
+		serviceConfig := new(confpb.Service)
+		fetchConfigUrl := util.FetchConfigURL(s.opts.ServiceManagementURL, s.serviceName, newConfigId)
+		if err := util.CallGooglelapis(s.client, fetchConfigUrl, util.GET, s.accessToken, serviceConfig); err != nil {
+			return "", nil, err
 		}
-		return nil, nil
+
+		return newRolloutId, serviceConfig, err
 	}
-	serviceConfig, err := _fetchConfig(configId)
-	if err == nil {
-		scf.curServiceConfig = serviceConfig
 
+	rolloutId, serviceConfig, err := _fetchConfig(configId)
+
+	if err == nil && serviceConfig != nil {
+		s.curRolloutId = rolloutId
+		s.curServiceConfig = serviceConfig
 	}
 
 	return serviceConfig, err
 }
 
-func (scf *ServiceConfigFetcher) SetFetchConfigTimer(interval *time.Duration, callback func(serviceConfig *confpb.Service)) {
+func (s *ServiceConfigFetcher) SetFetchConfigTimer(interval time.Duration, callback func(serviceConfig *confpb.Service)) {
 	go func() {
-		glog.Infof("start checking new rollouts every %v seconds", *interval)
-		scf.checkRolloutsTicker = time.NewTicker(*interval)
+		glog.Infof("start checking new rollouts every %v seconds", interval)
+		s.checkRolloutsTicker = time.NewTicker(interval)
 
-		for range scf.checkRolloutsTicker.C {
-			glog.Infof("check new rollouts for service %v", scf.serviceName)
+		for range s.checkRolloutsTicker.C {
+			glog.Infof("check new rollouts for service %v", s.serviceName)
 
-			serviceConfig, err := scf.FetchConfig("")
+			serviceConfig, err := s.FetchConfig("")
```

Comment:
I have not read FetchConfig() with empty yet.  But readability wise, I prefer the timer to call fetchLatestRolloutID() code
Or even better, the timer code is in the RolloutID fetcher. It will trigger ServiceConfigFetcher.

So the change to ServiceConfigFetcher should be smaller.